core/GlobalState.cc

```diff
@@ -2480,10 +2480,44 @@ MethodRef GlobalState::lookupStaticInitForClass(ClassOrModuleRef klass, bool all
     return ref;
 }
 
+namespace {
+
+template <typename GS> ClassOrModuleRef getStaticInitRootSingleton(GS &gs, FileRef file) {
+    auto &db = gs.packageDB();
+    if (!db.enabled()) {
+        return core::Symbols::rootSingleton();
+    }
+
+    // When we're processing package files, the package db isn't necessarily fully built yet, so we won't be able to
+    // reliably associate the file with a package. To avoid this problem, we stash all package file static-init methods
+    // on the rootSingleton.
+    if (file.data(gs).isPackage(gs)) {
+        return core::Symbols::rootSingleton();
+    }
+
+    auto pkgName = db.getPackageNameForFile(file);
+    if (!pkgName.exists()) {
+        return core::Symbols::rootSingleton();
+    }
+
+    if constexpr (std::is_const_v<GS>) {
+        auto root = pkgName.owner.data(gs)->lookupSingletonClass(gs);
```

Comment:
Is there a way to guarantee that the singleton class of a package's owner symbol is created when the attached class is created?

Then this helper would not need to be templated and could operate on a const GlobalState.